### PR TITLE
Store contract class data in client

### PR DESCRIPTION
**Pull request by tdelabro**

### CODE REVIEW

**Comment by tdelabro** [comment]
ci linter failed because counter10 doesn't exist on main, but it is built during the starknet-rpc-test and those passes. So not an issue

File: starknet-rpc-test/add_declare_transaction.rs [hunk 1]
```diff
@@ -145,13 +147,14 @@ async fn works_with_storage_change(madara: &ThreadSafeMadaraClient) -> Result<()
...
     }
-
-    assert!(rpc.get_class(BlockId::Number(block_number), expected_class_hash).await.is_ok());
-
     // included in block
     let included_txs = rpc.get_block_transaction_count(BlockId::Number(block_number)).await?;
     assert_eq!(included_txs, 1);
 
+    // Wait a bit to let the client sync the block
+    sleep(Duration::from_secs(3)).await;
```
Comment (EvolveArt) [inline starknet-rpc-test/add_declare_transaction.rs:155]: realistically how long does it take ?

Comment (EvolveArt) [reply starknet-rpc-test/add_declare_transaction.rs:155]: its not blocking for anything else than `get_class` right ?

File: starknet-rpc-test/get_class.rs [hunk 2]
```diff
@@ -47,58 +61,158 @@ async fn fail_non_existing_class_hash(madara: &ThreadSafeMadaraClient) -> Result
...
+                );
+
+                // decompress program
+                let mut gz = GzDecoder::new(&c.program[..]);
+                let mut decompressed_bytes = Vec::new();
+                gz.read_to_end(&mut decompressed_bytes).unwrap();
+                let mut program: LegacyProgram = serde_json::from_slice(decompressed_bytes.as_slice()).unwrap();
+
+                // Because of some obscure pathfinder bug, starknet-rs deserialize debug_info as an empty struct rather than none
+                // It has been fixed, so when https://github.com/xJonathanLEI/starknet-rs/pull/599 is megerd, this will become useless
```
Comment (EvolveArt) [inline starknet-rpc-test/get_class.rs:103]: merged no?

File: crates/client/contract-class-data/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,257 @@
+//! A worker to manage additional contact class data
+//!
+//! When receiving an `add_decalre_transaction` we convert the contract class we receive into a
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/client/contract-class-data/src/lib.rs:3]: done

File: crates/client/contract-class-data/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+{
+    // We push everything we get from the receiver here, then repetitively iterate over it to poll
+    // each transaction's status stream and take action when they end up being finalized.
+    let mut unfinalized_declare_transactions = Vec::new();
+    // Stores all the txs that were not yet mapped into our db when they finalized for later processing.
+    let mut finalized_but_not_mapped_yet = HashMap::<B::Hash, Vec<(TransactionHash, ClassHash)>>::new();
+    // Totally arbitrary limit.
+    // The job_recv being unbounded, there is not risk of it being full.
+    // Could be increased up to at most the max amount of declare tx possible to have in a single block.
+    // More that this would be useless as the "poll receiver" task is woke up at least one (but most
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/client/contract-class-data/src/lib.rs:125]: done

File: crates/client/contract-class-data/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+    let streams_buffer_limit = unsafe { NonZeroUsize::new_unchecked(10) };
+    let mut jobs_receiver = DeclareTransactionJobReceiverFuture::<B, P>::new(job_recv, streams_buffer_limit);
+
+    loop {
+        // Deal with new DeclareTx in the channel
+        {
+            // Check if new DeclareTx were created
+            let opt_amount = match poll!(jobs_receiver.next()) {
+                Poll::Pending => None,
+                Poll::Ready(None) => panic!("the ContractClassData sender has been droped"),
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/client/contract-class-data/src/lib.rs:136]: done

File: crates/client/contract-class-data/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+                            &mapping_db,
+                            transaction_hash,
+                            class_hash,
+                        );
+                    }
+
+                    // `swap_remove` for perfs, therefore don't increment `i`
+                    unfinalized_declare_transactions.swap_remove(i);
+                    // We did some blocking work, let's yield
+                    tokio::task::yield_now().await;
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/client/contract-class-data/src/lib.rs:202]: To do it more often, and not keep the tokio thread for ourselves too long.
We released it as soon as we did something a bit expensive.
In the unlikely situation where we have a really big vector, maybe we don't want to iter for the whole thing.

Maybe it's to much and we can be happy with yielding less ofter. I don't have a strong opinion on this

File: crates/client/contract-class-data/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+                | Poll::Ready(Some(TransactionStatus::Usurped(_)))
+                | Poll::Ready(Some(TransactionStatus::FinalityTimeout(_))) => {
+                    if let Err(e) = contract_class_db.remove_pending_contract_class_data(class_hash) {
+                        error!(
+                            "failed to remove pending contract class data for class with hashe
+            {class_hash:?}: {e}"
+                        )
+                    };
+                    unfinalized_declare_transactions.swap_remove(i);
+                    tokio::task::yield_now().await;
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/client/contract-class-data/src/lib.rs:218]: same

File: crates/client/contract-class-data/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+                _ => i += 1,
+            }
+        }
+
+        tokio::task::yield_now().await;
+    }
+}
+
+// To avoid false negative, this method should only be called after you made sure the Substrate
+// block containing the Starknet transaction with hash `transaction_hash` has been synced int
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/client/contract-class-data/src/lib.rs:231]: done

File: starknet-rpc-test/add_declare_transaction.rs (hunk 1 above)
Comment (tdelabro) [reply starknet-rpc-test/add_declare_transaction.rs:155]: I'm not sure. Probably less than 1sec.

Indeed, it's only an issue for `get_class` and `get_class_at`

File: starknet-rpc-test/get_class.rs (hunk 2 above)
Comment (tdelabro) [reply starknet-rpc-test/get_class.rs:103]: yup, it is now. But we will have to wait for a new release of starknet-rs

File: crates/client/contract-class-data/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+//! that the Starknet tx was successful. If yes, we move it to a more perenial storage, if not we
+//! remove it altogether from the temporary storage.
+//! The logic described above is splited beetween:
+//! - mc_rpc::add_declare_transaction
+//! - mc_db::ContractClassDataDb
+//! - mc_contract_class_data::run_worker (the present crate)
+//!
+//! The present worker does the following inside an infinite loop:
+//! - Poll the channel for new declare transactions and add them to a queue
+//! - Iterate over this queue and poll each individual tx watcher of update in the tx status.
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/client/contract-class-data/src/lib.rs:20]: done

File: crates/client/contract-class-data/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+//!  * If the Substrate tx failed in anyway, remove the contact class data.
+//!  * Else if it has been finalize check if the Substrate block has been synced in our backend.
+//!   + If it is not sync yet, moved it to another buffer, that will be checked again on each
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/client/contract-class-data/src/lib.rs:23]: done

File: crates/client/contract-class-data/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+        Self { receiver, buffer: Vec::with_capacity(limit.into()) }
+    }
+
+    #[inline(always)]
+    pub fn buffer_as_mut_ref(&mut self) -> &mut Vec<DeclareTransactionStatusStream<P>> {
+        &mut self.buffer
+    }
+}
+
+// When calling `poll_next`, if new values have been pushed to the channel, stores up to limit of
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/client/contract-class-data/src/lib.rs:84]: done

File: crates/client/contract-class-data/src/lib.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/client/contract-class-data/src/lib.rs:231]: done

File: crates/client/db/src/da_db.rs [hunk 12]
```diff
@@ -1,33 +1,24 @@
...
-                let felt = StarkFelt::decode(&mut &raw[..])?;
-                Ok(BlockHash(felt))
-            }
-            None => Err(DbError::ValueNotInitialized(
-                crate::columns::DA,
-                // Safe coze `LAST_PROVED_BLOCK` is utf8
-                unsafe { std::str::from_utf8_unchecked(crate::static_keys::LAST_PROVED_BLOCK) }.to_string(),
-            )),
-        }
+    pub fn last_proved_block(&self) -> Result<Option<BlockHash>, DbError> {
```
Comment (apoorvsadana) [inline crates/client/db/src/da_db.rs:14]: we can delete this actually, not used anymore. or we can create an issue if you want to do it in another PR :)

File: crates/client/rpc/src/contract_class_v0.rs [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+                .attributes
+                .into_iter()
+                .zip(accessible_scopes)
+                .map(|(attribute, accessible_scopes)| LegacyAttribute {
+                    accessible_scopes,
+                    end_pc: attribute.end_pc.try_into().expect("the value shoud have fit"),
+                    flow_tracking_data: attribute.flow_tracking_data.map(|ftd| LegacyFlowTrackingData {
+                        ap_tracking: LegacyApTrackingData {
+                            group: ftd.ap_tracking.group.try_into().expect("the value shoud have fit"),
+                            offset: ftd.ap_tracking.offset.try_into().expect("the value shoud have fit"),
```
Comment (apoorvsadana) [inline crates/client/rpc/src/contract_class_v0.rs:36]:
```suggestion
                    end_pc: attribute.end_pc.try_into().expect("the value should have fit"),
                    flow_tracking_data: attribute.flow_tracking_data.map(|ftd| LegacyFlowTrackingData {
                        ap_tracking: LegacyApTrackingData {
                            group: ftd.ap_tracking.group.try_into().expect("the value should have fit"),
                            offset: ftd.ap_tracking.offset.try_into().expect("the value should have fit"),
```

File: crates/primitives/transactions/src/from_broadcasted_transactions.rs [hunk 14]
```diff
@@ -85,8 +86,239 @@ pub fn try_account_tx_from_broadcasted_tx(
...
+        class_info: ClassInfo,
+        is_query: bool,
+    ) -> Result<DeclareTransaction, BroadcastedTransactionConversionError> {
+        if is_query {
+            DeclareTransaction::new_for_query(tx, tx_hash, class_info)
+        } else {
+            DeclareTransaction::new(tx, tx_hash, class_info)
+        }
+        .map_err(|_| BroadcastedTransactionConversionError::InvalidTransactionVersion)
+    }
```
Comment (apoorvsadana) [inline crates/primitives/transactions/src/from_broadcasted_transactions.rs:105]: this code is duplicated below. can move it to an outside function?

File: crates/primitives/transactions/src/from_broadcasted_transactions.rs [hunk 15]
```diff
@@ -85,8 +86,239 @@ pub fn try_account_tx_from_broadcasted_tx(
...
+
+    let user_tx = match broadcasted_declare_tx {
```
Comment (apoorvsadana) [inline crates/primitives/transactions/src/from_broadcasted_transactions.rs:107]: a lot of duplicate code here as well, can we move the common stuff to functions?

File: crates/client/contract-class-data/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (EvolveArt) [reply crates/client/contract-class-data/src/lib.rs:198]: I think it's not necessary to yield that often but yeah as you wish

**Review by EvolveArt — APPROVED** [review]

File: crates/node/src/service.rs [hunk 16]
```diff
@@ -369,6 +371,16 @@ pub fn new_full(
             }
         }
 
+        task_manager.spawn_essential_handle().spawn(
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/node/src/service.rs:374]: I asked internally about the future p2p protocol

File: crates/client/rpc/src/lib.rs [hunk 17]
```diff
@@ -292,40 +305,51 @@ where
...
+                error!("Failed store contract class data: {e}");
+                StarknetRpcApiError::InternalServerError
+            })?;
         }
+        let transaction_status_watcher_stream: Pin<Box<TransactionStatusStreamFor<P>>> =
+            submit_and_watch_extrinsic(self.pool.clone(), best_block_hash, extrinsic).await?;
+
+        self.contract_class_data_tx
+            .send((tx_hash, class_hash, transaction_status_watcher_stream))
+            .expect("this should work");
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/client/rpc/src/lib.rs:352]: done

File: crates/client/contract-class-data/src/lib.rs [hunk 18]
```diff
@@ -0,0 +1,257 @@
...
+    B: BlockT,
+    P: TransactionPool<Block = B>,
+{
+    // We push everything we get from the receiver here, then repetitively iterate over it to poll
+    // each transaction's status stream and take action when they end up being finalized.
+    let mut unfinalized_declare_transactions = Vec::new();
+    // Stores all the txs that were not yet mapped into our db when they finalized for later processing.
+    let mut finalized_but_not_mapped_yet = HashMap::<B::Hash, Vec<(TransactionHash, ClassHash)>>::new();
+    // Totally arbitrary limit.
+    // The job_recv being unbounded, there is not risk of it being full.
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/client/contract-class-data/src/lib.rs:123]: done

File: crates/client/contract-class-data/src/lib.rs [hunk 19]
```diff
@@ -0,0 +1,257 @@
...
+    // Could be increased up to at most the max amount of declare tx possible to have in a single block.
+    // More that this would be useless as the "poll receiver" task is woke up at least one (but most
+    // likely multiple times) during the span of one blocktime.
+    let streams_buffer_limit = unsafe { NonZeroUsize::new_unchecked(10) };
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/client/contract-class-data/src/lib.rs:127]: No need, we will just take 10 more in a later iteration of the loop. And there is no risk of congestion because the channel is unbounded.

Even in the hypothesis of blocks full of DeclareTx, I think we will still be able to empty the channel easily.

File: crates/client/contract-class-data/src/lib.rs [hunk 20]
```diff
@@ -0,0 +1,253 @@
...
+                    persist_or_remove_pending_contract_class_data(
+                        &contract_class_db,
+                        &mapping_db,
+                        *transaction_hash,
+                        *class_hash,
+                    );
+                }
+            }
+
+            // Don't retain the synced enties
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/client/contract-class-data/src/lib.rs:159]: done

File: crates/client/db/src/lib.rs [hunk 21]
```diff
@@ -69,6 +70,12 @@ pub(crate) mod columns {
 
     /// This column stores the fee paid on l1 for L1Handler transactions
     pub const L1_HANDLER_PAID_FEE: u32 = 7;
+
+    /// This column stores the additional data about contract classes we need to serve the RPCs
+    pub const CONTRACT_CLASS_DATA: u32 = 8;
+    /// This column stores data about classes that have been received by the node but are not yet
+    /// included in the chain
+    pub const PENDING_CONTRACT_CLASS_DATA: u32 = 9;
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/client/db/src/lib.rs:78]: So the risk is that when you kill your node you lose the mempool content.
Any tx that was there, will never be included in a block and never be moved out of PENDING.
So we could clean this up.
But there is the very unlikely possibility of a Tx being accepted in a block, but not being sync yet when the nodes shut down and therefore not being moved out.

So I would say that the cleanup logic has to first call `get_class` on the runtime for every entry in `PENDING` and only remove the ones that returned and error.

Wdyt?

[the remark this answers is not in the document]
Comment (apoorvsadana) [reply crates/client/db/src/lib.rs:78]: Yes, that makes sense!

File: crates/client/contract-class-data/src/lib.rs [hunk 22]
```diff
@@ -0,0 +1,253 @@
...
+    let mut finalized_but_not_mapped_yet = HashMap::<B::Hash, Vec<(TransactionHash, ClassHash)>>::new();
+    // Totally arbitrary limit.
+    // The job_recv being unbounded, there is not risk of it being full.
+    // Could be increased up to at most the max amount of declare tx possible to have in a single block.
+    // More that this would be useless as the "poll receiver" task is woke up at least once once but
+    // most likely multiple times) during the span of one blocktime.
+    let streams_buffer_limit = unsafe { NonZeroUsize::new_unchecked(10) };
+    let mut jobs_receiver = DeclareTransactionJobReceiverFuture::<B, P>::new(job_recv, streams_buffer_limit);
+
+    loop {
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/client/contract-class-data/src/lib.rs:130]: or what we can do is change the `read_contract_class_data` so that if it doesn't find it in the persisted storages, it goes to look for it in the other one.

[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/client/contract-class-data/src/lib.rs:130]: Better, always store in the persisted storage, and just stores the keys of pending in the pending one

[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/client/contract-class-data/src/lib.rs:130]: Now there is no need to wait

[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/client/contract-class-data/src/lib.rs:130]: Ah the issue, is that you now return a value for a TX that may have failed...
Did not solve anything...

[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/client/contract-class-data/src/lib.rs:130]: Solution: we remove them from PENDING during the mapping process.

[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/client/contract-class-data/src/lib.rs:130]: done

File: crates/client/db/src/contract_class_data_db.rs [hunk 23]
```diff
@@ -0,0 +1,105 @@
...
+
+use crate::{DbError, DbHash};
+
+/// Allow interaction with the mapping db
+pub struct ContractClassDataDb {
+    pub(crate) db: Arc<dyn Database<DbHash>>,
+}
+
+impl ContractClassDataDb {
+    pub fn register_pending_v0_contract_class_data(
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/client/db/src/contract_class_data_db.rs:16]: we need to know which type is being pushed in order to do this:
 ```
 data_as_json_vec.push(0);
 ```
 
 We would have to pass it as an extra argument, which feels unecessary.
 
 + a bit of duplicated code it's really not harmful here. We know for sure there is only 2 implementations

File: crates/client/db/src/da_db.rs (hunk 12 above)
Comment (tdelabro) [reply crates/client/db/src/da_db.rs:14]: We will start using it in the future, when we read messages from l1 again, won't we?

File: crates/client/rpc/src/contract_class_v0.rs (hunk 13 above)
Comment (tdelabro) [reply crates/client/rpc/src/contract_class_v0.rs]: done

File: crates/primitives/transactions/src/from_broadcasted_transactions.rs (hunk 14 above)
Comment (tdelabro) [reply crates/primitives/transactions/src/from_broadcasted_transactions.rs:105]: done

File: crates/primitives/transactions/src/from_broadcasted_transactions.rs (hunk 15 above)
Comment (tdelabro) [reply crates/primitives/transactions/src/from_broadcasted_transactions.rs:107]: I already did a bit of factorization here: https://github.com/keep-starknet-strange/madara/pull/1617

Let's merge #1617 first, then I will rebase this one, and do the factorization on top of it

File: crates/node/src/service.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/node/src/service.rs:374]: I think Starknet full nodes, relay the RPC calls directly, not the TXs.
So they will send the full data to the sequencer, it needs it to compile sierra into casm and check the hashes match

[the remark this answers is not in the document]
Comment (apoorvsadana) [reply crates/node/src/service.rs:374]: Oh ok, I thought we were talking about running Madara as a full node. That won't work anymore with this then, right?

File: crates/client/db/src/da_db.rs (hunk 12 above)
Comment (apoorvsadana) [reply crates/client/db/src/da_db.rs:14]: Hmm, I don't think we need the last proved block to read a message from the L1

File: crates/client/mapping-sync/src/sync_blocks.rs [hunk 24]
```diff
@@ -50,7 +51,28 @@ where
                             starknet_transaction_hashes: digest_starknet_block
                                 .transactions()
                                 .iter()
-                                .map(get_transaction_hash)
+                                .map(|tx| {
```
Comment (apoorvsadana) [inline crates/client/mapping-sync/src/sync_blocks.rs:54]: are the other changes necessary if we can do this? like we make no changes in the persistence of data but we just make sure that we persist during mapping

File: crates/node/src/service.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/node/src/service.rs:374]: Madara full node will also just relay the write RPC. not store anything

File: crates/client/db/src/lib.rs (hunk 21 above)
[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/client/db/src/lib.rs:78]: There is no easy way to implement this due to substrate abstraction layer over db + other stuffs.

Let's keep it as it is. Worst case we keep storing the classes of declare tx that were in the mempool when it was shut down. 

I'm fine with it for now.

File: crates/client/mapping-sync/src/sync_blocks.rs (hunk 24 above)
Comment (tdelabro) [reply crates/client/mapping-sync/src/sync_blocks.rs:54]: Declare are non revertible. If they fail they are not included into the block. But the substrate extrinsic will. This is what we are waiting for.

File: crates/node/src/service.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (apoorvsadana) [reply crates/node/src/service.rs:374]: Is it? I thought that only extrinsic are gossiped from the full node to the validator

[the remark this answers is not in the document]
Comment (apoorvsadana) [reply crates/node/src/service.rs:374]: Oh, I thought only extrinsics are gossiped from the full node to the validator in Substrate

**Comment by tdelabro** [comment]
repository archived in favor of https://github.com/madara-alliance/madara
